idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/proofs/inspections/coercionInspections.kt

```diff
@@ -0,0 +1,94 @@
+package arrow.meta.ide.plugins.proofs.inspections
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.ide.plugins.proofs.markers.coercionMessage
+import arrow.meta.ide.plugins.proofs.markers.participatingTypes
+import arrow.meta.phases.CompilerContext
+import arrow.meta.phases.Composite
+import arrow.meta.phases.ExtensionPhase
+import arrow.meta.plugins.proofs.phases.coerceProof
+import arrow.meta.quotes.ktFile
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.imports.importableFqName
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtValueArgument
+import org.jetbrains.kotlin.resolve.ImportPath
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.KotlinType
+
+val IdeMetaPlugin.coercionInspections: ExtensionPhase
+  get() = Composite(
+    implicitCoercionInspection,
+    explicitCoercionArgumentInspection,
+    explicitCoercionPropertyInspection
+  )
+
+internal fun KtDotQualifiedExpression.implicitParticipatingTypes(): Pair<KotlinType, KotlinType>? =
+  receiverExpression.resolveKotlinType().pairOrNull(selectorExpression?.resolveKotlinType())
+
+internal fun KtDotQualifiedExpression.coercionProofMessage(ctx: CompilerContext): String =
+  implicitParticipatingTypes()?.let { (subtype, supertype) ->
+    ctx.coerceProof(subtype, supertype)?.coercionMessage()
+  } ?: "Proof not found"
+
+internal fun KotlinType?.pairOrNull(b: KotlinType?): Pair<KotlinType, KotlinType>? =
+  if (this != null && b != null) Pair(this, b)
+  else null
+
+internal fun KtExpression.resolveKotlinType(): KotlinType? =
+  analyze(BodyResolveMode.PARTIAL).getType(this)
+
+internal fun KtElement.explicit(compilerContext: CompilerContext) =
+  when (this) {
+    is KtProperty -> explicit(compilerContext)
+    is KtValueArgument -> explicit(compilerContext)
+    else -> Unit
+  }
+
+private fun KtValueArgument.explicit(compilerContext: CompilerContext) {
```

Comment:
define this as an expression. same down below in the other 2 functions

Reply:
can you rephrase this?